### ISSUE TITLE
Failed micrograph error

### ISSUE THREAD

**Issue by danpf** [M4I-nanoscopy/InSilicoTEM#1]
Hello,
I think I have installed everything correctly (possibly not) but I'm getting the  following error when I run the RunTEMsim.m script:

Is there anything obvious that might be wrong here?
```
###################### Starting new Simulation ######################
       Micrographs: 1
       Size:        512x512
       Particles:   Randomized
       Defocus:     [300 - 800]nm
       Motion blur: 0.5
       Corr factor: 14
       Dose:        50e/A^2
                                 13-Jan-2021 23:22:21

Loading PDB file...
Elapsed time is 0.153397 seconds.
 
Generating random particle positions...
Warning: Max iteration reached.  Only 8 of the requested 22 circles drawn for radius 50.000 
> In Randomposition (line 148)
In RunTEMsim (line 97) 
Elapsed time is 0.009644 seconds.




-----------------------   Progress:   1 / 1 Micrographs   -----------------------
 
Particles:    8
Defocus:      400.2198nm

 
No need for generation of new particles. Load the potential maps from the folder 'Particles'...
./Raw/Particles/1dpx_a0001_Nx116_Ny114_Nz106_Alp42.2_Bet84.9_Gam343.0_MF0.5_VoxSize0.50A.raw
./Raw/Particles/1dpx_a0002_Nx110_Ny110_Nz118_Alp126.6_Bet106.4_Gam269.7_MF0.5_VoxSize0.50A.raw
./Raw/Particles/1dpx_a0003_Nx104_Ny130_Nz100_Alp291.0_Bet108.1_Gam29.5_MF0.5_VoxSize0.50A.raw
./Raw/Particles/1dpx_a0004_Nx106_Ny126_Nz110_Alp319.7_Bet153.9_Gam197.6_MF0.5_VoxSize0.50A.raw
./Raw/Particles/1dpx_a0005_Nx98_Ny114_Nz122_Alp212.7_Bet73.5_Gam335.8_MF0.5_VoxSize0.50A.raw
./Raw/Particles/1dpx_a0006_Nx104_Ny134_Nz104_Alp13.5_Bet36.7_Gam52.3_MF0.5_VoxSize0.50A.raw
./Raw/Particles/1dpx_a0007_Nx104_Ny104_Nz134_Alp235.8_Bet36.9_Gam355.5_MF0.5_VoxSize0.50A.raw
./Raw/Particles/1dpx_a0008_Nx124_Ny98_Nz112_Alp192.6_Bet79.5_Gam66.6_MF0.5_VoxSize0.50A.raw
Elapsed time is 0.489926 seconds.
Error using dip_image>matlabtype (line 2495)
Unknown data type string

Error in dip_image (line 246)
               [datatype,complex] = matlabtype(varargin{2});

Error in newim (line 74)
out = dip_image(0,dt);

Error in simTEM (line 42)
extprojstack  = newim(params2.proc.N,params2.proc.N,Nseries, 'complex');

Error in TEMsim (line 99)
[imStructOut] = simTEM(InputVol, params2);

Error in RunTEMsim (line 112)
                    out = TEMsim(defocus,mb_series(mb),cf_series(cf),1,rpdb,particles,dose(d),circles,pix,pixsize,pp,mindist);
 
```

**Comment by YZhang-ue** [comment]
Hi,
It seems that the problem is due to the **newim** function.
You can try to create a complex image with:
newim(512, 512, 1, 'complex');
after you successfully load the diplib.

If that does not work, it should be the diplip problem. You can try to reinstall the diplib properly.

Hope this helps.

**Comment by danpf** [comment]
hello
I have figured out the problem

the diplib library only has the datatypes "scomplex" or "dcomplex"
but some functions in this repo call "complex" which is not an option
they should be changed to scomplex or dcomplex

```
# grep -r "complex" *
extra/ReadImagic.m:% No complex data types are supported at present.
src/tiltingMS.m:%fix bug :rotation.m function doesn't work with complex input volume. Rotate 2 times (real and imag) or ask Bernd? 
src/multislice.m:psi_multislice = newim(Nm,Nm,'dcomplex')+1;
src/generateFullVolume.m:    VolPot = dip_image(permute(volstruct,[2 1 3]),'scomplex');  
src/generateFullVolume.m:    VolPot = dip_image(permute(volstruct,[2 1 3]),'scomplex');  
src/FreqDomRadon_Ewald_2d.m:stack = complex(stack,stack);
src/simulateCTF.m:    A = dip_image(ones(params2.proc.N,params2.proc.N),'complex');
src/simulateCTF.m:%         Image_com=dip_image(central_blocked_PP,'complex');
src/simulateCTF.m:        Image_com=dip_image(ones(params2.proc.N,params2.proc.N),'complex');
src/simulateCTF.m:        PhPlate = exp(1i*params2.mic.PP_Phase)*dip_image(ones(params2.proc.N,params2.proc.N),'complex');
src/simTEM.m:extprojstack  = newim(params2.proc.N,params2.proc.N,Nseries, 'complex');
src/simTEM.m:    psi_exit = newim(params2.proc.N, params2.proc.N, nTiltAngles, 'dcomplex');
src/simTEM.m:        psi_slice = newim(Nm,Nm,'scomplex')+1;
src/simTEM.m:%   Image_spp=dip_image(ones(params2.proc.N,params2.proc.N),'complex');
```

**Comment by danpf** [comment]
After changing some of the "complex" to dcomplex
I now get this error:
```
Unrecognized function or variable 'dip_fouriertransform'.

Error in simTEM (line 243)
  btot = ctf*dip_fouriertransform(dip_image(psi_exit(:,:,jjj)),'forward',[1 1 ]); %0]);

Error in TEMsim (line 99)
[imStructOut] = simTEM(InputVol, params2);

Error in RunTEMsim (line 112)
                    out = TEMsim(defocus,mb_series(mb),cf_series(cf),1,rpdb,particles,dose(d),circles,pix,pixsize,pp,mindist);
```
according to https://github.com/DIPlib/diplib/blob/e7dc2dfd4babab414f8bb6cc0d630fd017303f8b/changelogs/dipimage_3.0.beta.md

the function dip_fouriertransform has been renamed so that should be changed as well.

* additionally
the 'forward' should be changed to something else, because that option no longer exists.

**Comment by danpf** [comment]
Also:
```
Unrecognized function or variable 'WriteMRC'.

Error in RunTEMsim (line 119)
                    WriteMRC(double(ImageOut),0.50,[dir1 filesep char(s1)]);
 
```

There is no WriteMRC function provided by this library, only a ReadMRC

**Comment by danpf** [comment]
finally (sorry for so many comments)
my micrographs look like this which I don't think is correct -- do you think they are okay?
![image](https://user-images.githubusercontent.com/16297104/104773088-58cc6d00-5729-11eb-8395-23f0bf307cf8.png)

**Comment by YZhang-ue** [comment]
Hi Daniel，

No... it does not look good. Which diplib version are you using? I just
checked I am using dip/2.9. Probably different versions have function
differences.

**Comment by danpf** [comment]
where can i get version 2.9?
the oldest version listed on github is 3.0-beta1
https://github.com/DIPlib/diplib/tags

If you have a git sha associated with 2.9 i can maybe try  that

**Comment by YZhang-ue** [comment]
Here you go: https://diplib.org/download_2.9.html

**Comment by danpf** [comment]
Thank you @YZhang-ue ! Finally installed everything correctly!
The last error I am getting is:
```
Unrecognized function or variable 'WriteMRC'.
```
but now my images look nice!
![image](https://user-images.githubusercontent.com/16297104/105265949-f36fe600-5b45-11eb-886b-dc38493e9e90.png)

**Comment by danpf** [comment]
I have one final question though! 
Do you, or how do you, generate .star files for these micrographs so that relion can read them?

**Comment by YZhang-ue** [comment]
Congrats! 

I have uploaded two new scripts ( which can also be found online): WriteMRC and WriteMRCHeader.

For reading micrographs, Relion can read the .mrc files. 
.star file contains the information (coordinates, CTF, etc..) for each micrograph or particle. 
Note: You don't need to do motioncorr as there is no particle motion simulated.

Good luck!

**Comment by thorstenwagner** [comment]
I had the same problem and it was solved by using diplib 2.9

maybe its worth to mention that in the installation instructions
